### PR TITLE
FFmpegCutVideo postprocessor, for cutting the video after download

### PR DESCRIPTION
**Pull request by marnunez**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [X] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [ ] New extractor
- [X] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

This is a simple FFmpeg postprocessor for cutting videos at start-time and end-time:

`youtube-dl "URL" --start-time 10 --end-time 12`

There is a similar pull request for this feature that's nearly two years old and hasn't been worked for a long time now. This is understandable a hack: the whole video is downloaded for cutting just a segment. But I believe there is a need for this kind of postprocessing, and I've seen many a one-liner for this in the forums. It updates video duration if available, validates start-time and end-time and cuts only when needed, keeping all original stream codecs (-c copy). I believe this could do for the time being.

### CODE REVIEW

File: youtube_dl/__init__.py [hunk 1]
```diff
@@ -283,6 +283,13 @@ def parse_retries(retries):
         postprocessors.append({
             'key': 'FFmpegEmbedSubtitle',
         })
+    if opts.starttime or opts.endtime:
+        d = {'key': 'FFmpegCutVideo'}
+        if opts.starttime:
+            d['startTime'] = int(opts.starttime)
```
Comment (dstftw) [inline youtube_dl/__init__.py]: Incorrect case.

File: youtube_dl/postprocessor/ffmpeg.py [hunk 2]
```diff
@@ -215,6 +215,66 @@ def _ffmpeg_filename_argument(self, fn):
...
+            message += 'from %s ' % (start)
+            if duration:
+                duration -= self._startTime
+
+        if self._endTime:
+            end = self.toTime(self._endTime)
+            options.extend(['-to', end])
+            message += 'to %s' % (end)
+            if duration:
+                duration -= self._endTime
```
Comment (dstftw) [inline youtube_dl/postprocessor/ffmpeg.py]: Duration calculation is incorrect.

File: youtube_dl/postprocessor/ffmpeg.py [hunk 3]
```diff
@@ -215,6 +215,66 @@ def _ffmpeg_filename_argument(self, fn):
...
+
+        if '-to' not in options and '-ss' not in options:
+            self._downloader.to_screen('[ffmpeg] Nothing to cut. Keeping original')
+            return [], information
+
+        path = information['filepath']
+        temp_filename = prepend_extension(path, 'temp')
+        self._downloader.to_screen(message)
+        if duration:
+            information['duration'] = duration
```
Comment (dstftw) [inline youtube_dl/postprocessor/ffmpeg.py]: This has no effect. Postprocessors work on info dict copy.

**Comment by dstftw** [comment]
1. It must support cutting into multiple parts in order to be used for cutting by chapters.
2. It must not cut if start time is greater than duration.
3. It must not cut already downloaded files.

**Comment by marnunez** [comment]
I'm on it

File: youtube_dl/postprocessor/ffmpeg.py (hunk 3 above)
Comment (marnunez) [reply youtube_dl/postprocessor/ffmpeg.py]: I've seen this used [here](https://github.com/marnunez/youtube-dl/blob/afc763ee2e193b5711e9af53b1e133031611e72d/youtube_dl/postprocessor/ffmpeg.py#L299) and [here](https://github.com/marnunez/youtube-dl/blob/afc763ee2e193b5711e9af53b1e133031611e72d/youtube_dl/postprocessor/ffmpeg.py#L354) with filepath and other keys. Does this work in this cases?

**Comment by Aniruddh-J** [comment]
And maybe force downloader to `ffmpeg` if an external downloader is defined in config?

**Comment by mohseneslami** [comment]
It's not working for oreilly working. please fix it.

oreilly change login url members

**Pull request closed without merging**